src/position.rs

```diff
@@ -560,45 +564,52 @@ impl Position {
             MoveRecord::Normal {
                 from,
                 to,
-                ref moved,
+                ref placed,
                 ref captured,
                 promoted,
             } => {
                 if *self.piece_at(from) != None {
-                    return Err(MoveError::Inconsistent);
+                    return Err(MoveError::Inconsistent(
+                        "`from` of the move is filled by another piece",
+                    ));
                 }
 
-                let pc = if promoted {
-                    match moved.unpromote() {
+                let moved = if promoted {
+                    match placed.unpromote() {
                         Some(unpromoted) => unpromoted,
-                        None => return Err(MoveError::Inconsistent),
+                        None => return Err(MoveError::Inconsistent("Cannot unpromoted the piece")),
                     }
                 } else {
-                    *moved
+                    *placed
                 };
-                if *self.piece_at(to) != Some(pc) {
-                    return Err(MoveError::Inconsistent);
+                if *self.piece_at(to) != Some(*placed) {
+                    return Err(MoveError::Inconsistent(
+                        "Expected piece is not found in `to`",
+                    ));
                 }
 
-                self.set_piece(from, Some(*moved));
+                self.set_piece(from, Some(moved));
                 self.set_piece(to, *captured);
                 self.occupied_bb ^= from;
                 self.occupied_bb ^= to;
                 self.type_bb[moved.piece_type.index()] ^= from;
-                self.type_bb[moved.piece_type.index()] ^= to;
+                self.type_bb[placed.piece_type.index()] ^= to;
                 self.color_bb[moved.color.index()] ^= from;
-                self.color_bb[moved.color.index()] ^= to;
+                self.color_bb[placed.color.index()] ^= to;
 
                 if let Some(ref cap) = *captured {
                     self.occupied_bb ^= to;
                     self.type_bb[cap.piece_type.index()] ^= to;
                     self.color_bb[cap.color.index()] ^= to;
-                    self.hand.decrement(cap.flip());
+                    let unpromoted_cap = cap.unpromote().unwrap_or(*cap);
+                    self.hand.decrement(unpromoted_cap.flip());
```

Comment:
This is a fix for the first item. The rest is for the second and also for enriching error messages.

Reply:
Ah, this does make sense :tada: